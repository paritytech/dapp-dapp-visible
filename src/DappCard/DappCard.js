// Copyright 2015-2017 Parity Technologies (UK) Ltd.
// This file is part of Parity.

// Parity is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.

// Parity is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.

// You should have received a copy of the GNU General Public License
// along with Parity.  If not, see <http://www.gnu.org/licenses/>.

import React, { Component } from 'react';
import { observer } from 'mobx-react';
import PropTypes from 'prop-types';
import Card from 'semantic-ui-react/dist/commonjs/views/Card';
import Button from 'semantic-ui-react/dist/commonjs/elements/Button';
import Label from 'semantic-ui-react/dist/commonjs/elements/Label';
import DappIcon from '@parity/ui/lib/DappIcon';
import { FormattedMessage } from 'react-intl';

import DappVouchFor from './DappVouchFor';
import styles from './DappCard.css';

const DAPP_DAPP_VISIBLE_ID =
  '0xa48bd8fd56c90c899135281967a6cf90865c221b46f27f9fbe2a236d74a64ea2'; // This dapp's id

class DappCard extends Component {
  static propTypes = {
    dapp: PropTypes.shape({
      id: PropTypes.string.isRequired,
      name: PropTypes.string.isRequired,
      image: PropTypes.string
    }).isRequired,
    onAdd: PropTypes.func.isRequired,
    onOpen: PropTypes.func.isRequired,
    onRemove: PropTypes.func.isRequired,
    visible: PropTypes.bool
  };

  handleAdd = () => this.props.onAdd(this.props.dapp.id);

  handleRemove = () => this.props.onRemove(this.props.dapp.id);

  handleOpen = () => this.props.onOpen(this.props.dapp.id);

  render() {
    const { dapp, visible } = this.props;

    return (
      <Card>
        <Card.Content>
          <DappIcon
            app={dapp}
            className={`ui centered image ${styles.picture}`}
          />
          <Card.Header>{dapp.name}</Card.Header>
          <Card.Meta>{dapp.description}</Card.Meta>
          <Card.Description>
            {dapp.type === 'local' && (
              <Label>
                <FormattedMessage
                  id="dapp.visible.localLabel"
                  defaultMessage="Local Dapp"
                />
              </Label>
            )}
            {dapp.type === 'builtin' && (
              <Label>
                <FormattedMessage
                  id="dapp.visible.builtinLabel"
                  defaultMessage="By Parity"
                />
              </Label>
            )}
            <DappVouchFor dapp={dapp} />
          </Card.Description>
        </Card.Content>
        <Card.Content extra>
          <Button
            basic
            secondary
            content={
              <FormattedMessage
                id="dapp.visible.openDapp"
                defaultMessage="Open"
              />
            }
            icon="external"
            onClick={this.handleOpen}
          />
          {visible ? (
            <Button
              content={
                <FormattedMessage
                  id="dapp.visible.addedDapp"
                  defaultMessage="Added to home"
                />
              }
              disabled={dapp.id === DAPP_DAPP_VISIBLE_ID} // Do not allow removing this dapp from homepage
              icon="toggle on"
              onClick={this.handleRemove}
              primary
            />
          ) : (
            <Button
              basic
              content={
                <FormattedMessage
                  id="dapp.visible.addDapp"
                  defaultMessage="Add to home"
                />
              }
              icon="toggle off"
              onClick={this.handleAdd}
              primary
            />
          )}
        </Card.Content>
      </Card>
    );
  }
}

export default observer(DappCard);
