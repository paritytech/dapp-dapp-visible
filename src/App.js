// Copyright 2015-2017 Parity Technologies (UK) Ltd.
// This file is part of Parity.

// Parity is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.

// Parity is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.

// You should have received a copy of the GNU General Public License
// along with Parity.  If not, see <http://www.gnu.org/licenses/>.

import { observer } from 'mobx-react';
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { FormattedMessage } from 'react-intl';

import Card from 'semantic-ui-react/dist/commonjs/views/Card';
import Header from 'semantic-ui-react/dist/commonjs/elements/Header';
import ActionBar from '@parity/ui/lib/Actionbar/actionbar';

import DappCard from './DappCard';
import styles from './App.css';

class App extends Component {
  handleToggleVisibility = dappId => {
    const { store } = this.props;
    if (store.displayApps[dappId]) {
      store.hideApp(dappId);
    } else {
      store.showApp(dappId);
    }
  };

  renderList = (title, subtitle, items) => {
    const { store } = this.props;
    if (!items || !items.length) {
      return null;
    }

    return (
      <div className={styles.sectionList}>
        <Header as="h4">{title}</Header>
        <div>{subtitle}</div>
        <Card.Group stackable className={styles.cardGroup}>
          {items.map((dapp, index) => (
            <DappCard
              key={index}
              dapp={dapp}
              visible={store.displayApps[dapp.id]}
              onClick={() => this.handleToggleVisibility(dapp.id)}
            />
          ))}
        </Card.Group>
      </div>
    );
  };

  render() {
    const { store } = this.props;
    console.log(store.apps);
    return (
      <div className={styles.layout}>
        <ActionBar
          title={
            <FormattedMessage
              id="dapps.visible.title"
              defaultMessage="Visible Dapps"
            />
          }
        />
        {this.renderList(
          <FormattedMessage
            id="dapps.visible.local.sectionTitle"
            defaultMessage="Applications locally available"
          />,
          <FormattedMessage
            id="dapps.visible.local.sectionSubtitle"
            defaultMessage="All applications installed locally on the machine by the user for access by the Parity client."
          />,
          store.sortedLocal
        )}
        {this.renderList(
          <FormattedMessage
            id="dapps.visible.builtin.sectionTitle"
            defaultMessage="Applications bundled with Parity"
          />,
          <FormattedMessage
            id="dapps.visible.builtin.sectionSubtitle"
            defaultMessage="Experimental applications developed by the Parity team to show off dapp capabilities, integration, experimental features and to control certain network-wide client behaviour."
          />,
          store.sortedBuiltin
        )}
        {this.renderList(
          <FormattedMessage
            id="dapps.visible.network.sectionTitle"
            defaultMessage="Applications on the global network"
          />,
          <FormattedMessage
            id="dapps.visible.network.sectionSubtitle"
            defaultMessage="These applications are not affiliated with Parity nor are they published by Parity. Each remain under the control of their respective authors. Please ensure that you understand the goals for each application before interacting."
          />,
          store.sortedNetwork
        )}
      </div>
    );
  }
}

export default observer(App);

App.propTypes = {
  store: PropTypes.object.isRequired
};
