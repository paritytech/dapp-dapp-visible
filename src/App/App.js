// Copyright 2015-2017 Parity Technologies (UK) Ltd.
// This file is part of Parity.

// Parity is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.

// Parity is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.

// You should have received a copy of the GNU General Public License
// along with Parity.  If not, see <http://www.gnu.org/licenses/>.

import React, { Component } from 'react';
import { observer } from 'mobx-react';
import PropTypes from 'prop-types';
import { FormattedMessage, injectIntl, intlShape } from 'react-intl';

import Page from '@parity/ui/lib/Page';
import Grid from 'semantic-ui-react/dist/commonjs/collections/Grid';
import Menu from 'semantic-ui-react/dist/commonjs/collections/Menu';
import Input from 'semantic-ui-react/dist/commonjs/elements/Input';
import Header from 'semantic-ui-react/dist/commonjs/elements/Header';
import Label from 'semantic-ui-react/dist/commonjs/elements/Label';
import Card from 'semantic-ui-react/dist/commonjs/views/Card';

import DappCard from '../DappCard';
import DappsStore from './store';
import styles from './App.css';

const ALL_DAPPS = 'ALL_DAPPS';
const MY_DAPPS = 'MY_DAPPS';

export class App extends Component {
  static contextTypes = {
    api: PropTypes.object.isRequired
  };

  static propTypes = {
    intl: intlShape
  };

  state = {
    menu: ALL_DAPPS, // or MY_DAPPS
    value: ''
  };

  dappsStore = new DappsStore(this.context.api);

  handleMenuClick = (_, { name }) => this.setState({ menu: name });

  handleSearch = (_, { value }) => this.setState({ value });

  renderList = (title, subtitle, items) => {
    const { intl: { formatMessage } } = this.props;
    return (
      <div>
        <Header as="h4">{title}</Header>
        {subtitle}
        <br />
        <Input
          fluid
          iconPosition="left"
          icon="search"
          placeholder={formatMessage({
            id: 'dapps.visible.search',
            defaultMessage: 'Search for dapps by name, description...'
          })}
          className={styles.search}
          onChange={this.handleSearch}
        />
        {items && (
          <Card.Group stackable className={styles.cardGroup}>
            {items
              .filter(
                ({ name, description }) =>
                  this.state.value
                    ? name.includes(this.state.value) ||
                      description.includes(this.state.value)
                    : true
              )
              .map((dapp, index) => (
                <DappCard
                  key={`${dapp.id}-${index}`}
                  dapp={dapp}
                  visible={
                    this.dappsStore.displayApps[dapp.id] &&
                    this.dappsStore.displayApps[dapp.id].visible
                  }
                  onAdd={this.dappsStore.showApp}
                  onRemove={this.dappsStore.hideApp}
                  onOpen={this.dappsStore.loadApp}
                />
              ))}
          </Card.Group>
        )}
      </div>
    );
  };

  render() {
    return (
      <Page
        title={
          <FormattedMessage
            id="dapps.visible.title"
            defaultMessage="Browse Dapps"
          />
        }
      >
        <Grid>
          <Grid.Row>
            <Grid.Column width={4}>
              <Menu pointing secondary vertical>
                <Menu.Item
                  name={ALL_DAPPS}
                  active={this.state.menu === ALL_DAPPS}
                  onClick={this.handleMenuClick}
                >
                  <FormattedMessage
                    id="dapps.visible.allDapps"
                    defaultMessage="All Dapps"
                  />
                </Menu.Item>
                <Menu.Item
                  name={MY_DAPPS}
                  active={this.state.menu === MY_DAPPS}
                  onClick={this.handleMenuClick}
                >
                  <FormattedMessage
                    id="dapps.visible.myDapps"
                    defaultMessage="My Dapps"
                  />
                  <Label>{this.dappsStore.visibleApps.length}</Label>
                </Menu.Item>
              </Menu>
            </Grid.Column>
            <Grid.Column width={12}>
              {this.state.menu === ALL_DAPPS
                ? this.renderList(
                    <FormattedMessage
                      id="dapps.visible.local.sectionTitleAllDapps"
                      defaultMessage="All available Dapps"
                    />,
                    <FormattedMessage
                      id="dapps.visible.local.sectionSubtitleAllDapps"
                      defaultMessage="Here are all available dapps for your Parity Client. You can add them on your homepage or try them out here."
                    />,
                    this.dappsStore.apps
                  )
                : this.renderList(
                    <FormattedMessage
                      id="dapps.visible.local.sectionTitleMyDapps"
                      defaultMessage="My Applications"
                    />,
                    <FormattedMessage
                      id="dapps.visible.local.sectionSubtitleMyDapps"
                      defaultMessage="The dapps here are added on your homepage for quick access."
                    />,
                    this.dappsStore.visibleApps
                  )}
            </Grid.Column>
          </Grid.Row>
        </Grid>
      </Page>
    );
  }
}

export default injectIntl(observer(App));
