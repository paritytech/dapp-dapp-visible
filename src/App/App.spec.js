// Copyright 2015-2017 Parity Technologies (UK) Ltd.
// This file is part of Parity.

// Parity is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.

// Parity is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.

// You should have received a copy of the GNU General Public License
// along with Parity.  If not, see <http://www.gnu.org/licenses/>.

import React from 'react';
import { shallowToJson } from 'enzyme-to-json';
import Menu from 'semantic-ui-react/dist/commonjs/collections/Menu';
import Input from 'semantic-ui-react/dist/commonjs/elements/Input';

import { shallowWithIntl } from '../setupTests';
import DappCard from '../DappCard';
import DappsStore from './store';
import { App } from './App';

jest.mock('./store');
const mockStore = {
  apps: [
    { id: '123', name: '123', description: '' },
    { id: '456', name: '456', description: '' }
  ],
  visibleApps: [{ id: '456', name: '456', description: '' }],
  displayApps: { '123': { visible: false }, '456': { visible: true } },
  showApp: () => {},
  hideApp: () => {},
  loadApp: () => {}
};

test('should render correctly when we display all apps', () => {
  DappsStore.mockImplementation(() => mockStore);
  const component = shallowWithIntl(<App />, { context: { api: {} } });

  expect(shallowToJson(component)).toMatchSnapshot();
});

test('should render correctly when we display my apps', () => {
  DappsStore.mockImplementation(() => mockStore);
  const component = shallowWithIntl(<App />, { context: { api: {} } });
  component.setState({ menu: 'MY_DAPPS' });

  expect(shallowToJson(component)).toMatchSnapshot();
});

test('should handle correctly changing menus', () => {
  DappsStore.mockImplementation(() => mockStore);
  const component = shallowWithIntl(<App />, { context: { api: {} } });

  component
    .find(Menu.Item)
    .at(1)
    .simulate('click', null, { name: 'MY_DAPPS' });
  expect(component.state('menu')).toBe('MY_DAPPS');

  component
    .find(Menu.Item)
    .at(0)
    .simulate('click', null, { name: 'ALL_DAPPS' });
  expect(component.state('menu')).toBe('ALL_DAPPS');
});

test('should handle searching dapps', () => {
  DappsStore.mockImplementation(() => mockStore);
  const component = shallowWithIntl(<App />, { context: { api: {} } });

  component
    .find(Input)
    .first()
    .props()
    .onChange(null, { value: '12' });
  expect(component.state('value')).toBe('12');
  component.update();
  expect(component.find('DappCard').length).toBe(1);
});

test('should call store.loadApp when we click on open App', () => {
  const loadApp = jest.fn();
  DappsStore.mockImplementation(() => ({ ...mockStore, loadApp }));
  const component = shallowWithIntl(<App />, { context: { api: {} } });

  component
    .find(DappCard)
    .first() // 123
    .props()
    .onOpen();

  expect(loadApp).toHaveBeenCalled();
});

test('should call store.showApp when we add to home', () => {
  const showApp = jest.fn();
  DappsStore.mockImplementation(() => ({ ...mockStore, showApp }));
  const component = shallowWithIntl(<App />, { context: { api: {} } });

  component
    .find(DappCard)
    .first() // 123
    .props()
    .onAdd();

  expect(showApp).toHaveBeenCalled();
});

test('should call store.hideApp when we hide it', () => {
  const hideApp = jest.fn();
  DappsStore.mockImplementation(() => ({ ...mockStore, hideApp }));
  const component = shallowWithIntl(<App />, { context: { api: {} } });

  component
    .find(DappCard)
    .last() // 456
    .props()
    .onRemove();

  expect(hideApp).toHaveBeenCalled();
});
