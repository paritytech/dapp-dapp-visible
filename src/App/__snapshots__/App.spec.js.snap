// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly when we display all apps 1`] = `
<Page
  title={
    <FormattedMessage
      defaultMessage="Browse Dapps"
      id="dapps.visible.title"
      values={Object {}}
    />
  }
>
  <Grid>
    <GridRow>
      <GridColumn
        width={4}
      >
        <Menu
          pointing={true}
          secondary={true}
          vertical={true}
        >
          <MenuItem
            active={true}
            name="ALL_DAPPS"
            onClick={[Function]}
          >
            <FormattedMessage
              defaultMessage="All Dapps"
              id="dapps.visible.allDapps"
              values={Object {}}
            />
          </MenuItem>
          <MenuItem
            active={false}
            name="MY_DAPPS"
            onClick={[Function]}
          >
            <FormattedMessage
              defaultMessage="My Dapps"
              id="dapps.visible.myDapps"
              values={Object {}}
            />
            <Label>
              1
            </Label>
          </MenuItem>
        </Menu>
      </GridColumn>
      <GridColumn
        width={12}
      >
        <div>
          <Header
            as="h4"
          >
            <FormattedMessage
              defaultMessage="All available Dapps"
              id="dapps.visible.local.sectionTitleAllDapps"
              values={Object {}}
            />
          </Header>
          <FormattedMessage
            defaultMessage="Here are all available dapps for your Parity Client. You can add them on your homepage or try them out here."
            id="dapps.visible.local.sectionSubtitleAllDapps"
            values={Object {}}
          />
          <br />
          <Input
            fluid={true}
            icon="search"
            iconPosition="left"
            onChange={[Function]}
            placeholder="Search for dapps by name, description..."
            type="text"
          />
          <CardGroup
            stackable={true}
          >
            <DappCard
              dapp={
                Object {
                  "description": "",
                  "id": "123",
                  "name": "123",
                }
              }
              key="123-0"
              onAdd={[Function]}
              onOpen={[Function]}
              onRemove={[Function]}
              visible={false}
            />
            <DappCard
              dapp={
                Object {
                  "description": "",
                  "id": "456",
                  "name": "456",
                }
              }
              key="456-1"
              onAdd={[Function]}
              onOpen={[Function]}
              onRemove={[Function]}
              visible={true}
            />
          </CardGroup>
        </div>
      </GridColumn>
    </GridRow>
  </Grid>
</Page>
`;

exports[`should render correctly when we display my apps 1`] = `
<Page
  title={
    <FormattedMessage
      defaultMessage="Browse Dapps"
      id="dapps.visible.title"
      values={Object {}}
    />
  }
>
  <Grid>
    <GridRow>
      <GridColumn
        width={4}
      >
        <Menu
          pointing={true}
          secondary={true}
          vertical={true}
        >
          <MenuItem
            active={false}
            name="ALL_DAPPS"
            onClick={[Function]}
          >
            <FormattedMessage
              defaultMessage="All Dapps"
              id="dapps.visible.allDapps"
              values={Object {}}
            />
          </MenuItem>
          <MenuItem
            active={true}
            name="MY_DAPPS"
            onClick={[Function]}
          >
            <FormattedMessage
              defaultMessage="My Dapps"
              id="dapps.visible.myDapps"
              values={Object {}}
            />
            <Label>
              1
            </Label>
          </MenuItem>
        </Menu>
      </GridColumn>
      <GridColumn
        width={12}
      >
        <div>
          <Header
            as="h4"
          >
            <FormattedMessage
              defaultMessage="My Applications"
              id="dapps.visible.local.sectionTitleMyDapps"
              values={Object {}}
            />
          </Header>
          <FormattedMessage
            defaultMessage="The dapps here are added on your homepage for quick access."
            id="dapps.visible.local.sectionSubtitleMyDapps"
            values={Object {}}
          />
          <br />
          <Input
            fluid={true}
            icon="search"
            iconPosition="left"
            onChange={[Function]}
            placeholder="Search for dapps by name, description..."
            type="text"
          />
          <CardGroup
            stackable={true}
          >
            <DappCard
              dapp={
                Object {
                  "description": "",
                  "id": "456",
                  "name": "456",
                }
              }
              key="456-0"
              onAdd={[Function]}
              onOpen={[Function]}
              onRemove={[Function]}
              visible={true}
            />
          </CardGroup>
        </div>
      </GridColumn>
    </GridRow>
  </Grid>
</Page>
`;
